Add a Physics class for linearFV turbulence modelling

GiudGiud opened the pull request:
refs https://github.com/idaholab/moose/issues/29175

Testing includes the traditional other Physics we like to use with turbulence

GiudGiud committed:
Factor out common code in turbulence Physics into a base class
refs #29175

GiudGiud committed:
Add missing parameter groups
Register linearFV turbulence physics
refs #29175

GiudGiud committed:
Add a linear FV version of the turbulence physics:
- only k-epsilon implemented
refs #29175

GiudGiud committed:
Remove unused linearized model parameter from TKE/D source sink kernels

GiudGiud committed:
Forward C_mu parameter from Physics to kernels and BCs

GiudGiud committed:
Update doco for current status of Physics support
refs #29175

GiudGiud committed:
Add test for turbulence case with linearFV + physics
Factor out common parts of the input
refs #29175

GiudGiud committed:
Adapt auxvariable quantities definitions for linearFV k-epsilon
Add missing auxkernel definitions for k-epsilon within turbulence Physics syntax
refs #29175

GiudGiud committed:
Use the effective viscosity for the advection kernel when in presence of a turbulence model
Rename orthogonality correction parameter for convenience
refs #29175

GiudGiud committed:
Add a good default for system parameter for linearFV turbulence physics
Add missing orthogonality correction parameter
Remove advection kernel for momentum since it will be done in FlowPhysics
Change the variable type to linear
Remove now-removed parameters from source & sink kernels
refs #29175

GiudGiud committed:
Gold this test independently since the Physics includes the deviatoric term
for dynamic viscosity, while the regular input only includes it for turbulent viscosity

GiudGiud committed:
Add test for the heated turbulent channel
with T and h variables

GiudGiud committed:
Add one more Block restriction API on Physics
We really need an interface for Actions with block restriction

GiudGiud committed:
Move turbulent heat diffusion kernels to fluid heat transfer physics for linearFV
Add support for enthalpy variable
Add support for block restriction of diffusion

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3174011/), step Generate coverage on 9448a86 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>b36d43</th>
      <th colspan=4>#31229 9448a8</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.67%</td>
      <td>85.67%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>117988</td>
      <td>117988</td>
      <td>-</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19728</td>
      <td>19728</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31229/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31229/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>b36d43</th>
      <th colspan=4>#31229 9448a8</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.49%</td>
      <td>85.42%</td>
      <td>-0.08%</td>
      <td>85.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>19847</td>
      <td>20031</td>
      <td>+184</td>
      <td>442</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3368</td>
      <td>3420</td>
      <td>+52</td>
      <td>78</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31229/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31229/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/31229/coverage/xfem)

</details>

## Warnings

- `navier_stokes` new line coverage rate 85.00% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud committed:
Add test for passive scalar advection and turbulence and linearFV

GiudGiud committed:
Add documentation for the turbulence Physics

GiudGiud committed:
Misc test and non-unity build fixes

GiudGiud committed:
Update table for coupling cases we have tried
They are not all under CI..

grmnptr commented on modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C:
```diff
@@ -0,0 +1,345 @@
+{
+  InputParameters params = WCNSFVTurbulencePhysicsBase::validParams();
+  params.addClassDescription(
+      "Define a turbulence model for a incompressible or weakly-compressible Navier Stokes "
```
```suggestion
      "Define a turbulence model for an incompressible or weakly-compressible Navier Stokes "
```

grmnptr commented on modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C:
```diff
@@ -0,0 +1,345 @@
+    // Simpler to add the kernel in the fluid heat transfer physics with k + k_t instead of two
+    // kernels one with k, one with k_t
+  }
+}
```
```suggestion
```

grmnptr commented on modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C:
```diff
@@ -0,0 +1,345 @@
+    params.set<LinearVariableName>("variable") = _tked_name;
+    params.set<MooseFunctorName>("scaling_coeff") = getParam<MooseFunctorName>("sigma_eps");
+    getProblem().addLinearFVKernel(kernel_type, prefix() + "tked_diffusion_mu_turb", params);
+  }
```
We should have only one block `mu + mu_t/sigma`. Linear FV doesn't work very well with the definition of multiple kernels for the same physics. If you define a flux BC it will be added twice because of this.

GiudGiud replied:
so if we use an auxkernel for mu_eff = mu_t + mu I think we would not use the boundary condition

GiudGiud replied:
ok so we go the functor material route

GiudGiud replied:
all 3. a little bit

grmnptr reviewed (changes requested):
Looks nice, just a few questions!

GiudGiud committed:
Address Peter's review

- better doco
- add errors and suppressions on unimplemented features for k-epsilon
- add rho multiplier on the time derivatives

Co-authored-by: Peter German <31662443+grmnptr@users.noreply.github.com>

GiudGiud commented on modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C:
```diff
@@ -0,0 +1,514 @@
+      "tke_two_term_bc_expansion",
+      false,
+      "If a two-term Taylor expansion is needed for the determination of the boundary values"
+      "of the turbulent kinetic energy.");
```
already suppressed
  params.suppressParameter<bool>("tke_two_term_bc_expansion");
  params.suppressParameter<bool>("tked_two_term_bc_expansion");

GiudGiud commented on modules/navier_stokes/src/physics/WCNSFVTurbulencePhysicsBase.C:
```diff
@@ -0,0 +1,514 @@
+    }
+    if (!getParam<bool>("mu_t_as_aux_variable"))
+    {
+      // TODO: suppress this for linearFV
```
already done

GiudGiud commented on modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C:
```diff
@@ -0,0 +1,345 @@
+WCNSLinearFVTurbulencePhysics::addSolverVariables()
+{
+  if (_turbulence_model == "mixing-length" || _turbulence_model == "none")
+    return;
```
added one

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/ins/turbulence/channel/linear-segregated/gold/channel_ERCOFTAC-physics_out_line_center_channel_0001.csv:
```diff
@@ -0,0 +1,5 @@
+0.0025877697461348,7.3889023714783e-05,0,0.34151550458492,1.040915708239,-0.00060794749797394,12.5,0.0001,0
+0.0022716271031286,0.00010389410436847,25,0.19805242257515,1.1031328293788,-0.00018981906625752,37.5,0.0001,0
+0.0024687725070957,0.00013918985689074,50,0.11414494135339,1.1159091029636,-9.0708639485251e-05,62.5,0.0001,0
+0.0027809910727729,0.00016717995137461,75,0.03957707667781,1.1166179183553,-2.7900067360663e-05,87.5,0.0001,0
```
we would need to add a knob to turn off the deviatoric term

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3311721/), step Results summary on cfbc15c wanted to post the following:

## Framework test summary

### Removed tests
None
### Added tests
None
### Run time changes
| Test                                                             |   Base (s) |   Head (s) | +/-     |
|------------------------------------------------------------------|------------|------------|---------|
| `misc/max_var_n_dofs_per_elem.1d`                                |       0.84 |       1.34 | +59.63% |
| `meshgenerators/batch_mesh_generation_action.double_scalar_corr` |       0.82 |       1.29 | +56.51% |
## Modules test summary

### Removed tests
None
### Added tests
| Test                                                                                                                        |   Time (s) |
|-----------------------------------------------------------------------------------------------------------------------------|------------|
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_heated_physics`          |       4.5  |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_heated_enthalpy_physics` |       4.44 |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_scalar_physics`          |       4.15 |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_physics`                 |       4.14 |
### Run time changes
| Test                                                |   Base (s) |   Head (s) | +/-     |
|-----------------------------------------------------|------------|------------|---------|
| `misc/test:sensor_postprocessor.time_delay_test`    |       3.03 |       5.08 | +67.87% |
| `solid_mechanics/test:drucker_prager.small2_native` |       1.5  |       2.4  | +59.81% |

GiudGiud committed:
Add a functor to compute the turbulent dynamic viscosity

GiudGiud committed:
Add functor materials for each turbulent viscosity used in the various diffusion kernels

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3311946/), step Results summary on 2be670c wanted to post the following:

## Framework test summary

### Removed tests
None
### Added tests
None
### Run time changes
| Test                                                                            |   Base (s) |   Head (s) | +/-     |
|---------------------------------------------------------------------------------|------------|------------|---------|
| `fvkernels/fv_burgers.fv_burgers`                                               |       0.92 |       1.56 | +68.80% |
| `meshgenerators/stitch_mesh_generator.stitched_mesh_generator/overlapping_bids` |       0.82 |       1.37 | +66.10% |
| `meshgenerators/polyline_mesh_generator.polyline_good_boundary`                 |       0.8  |       1.28 | +60.33% |
| `tag.not_zeroed_vector_tag`                                                     |       0.71 |       1.13 | +60.03% |
| `fvkernels/two-var-flux-and-kernel.jac`                                         |       0.9  |       1.43 | +59.36% |
| `postprocessors/interface_diffusive_flux.interface_diffusive_flux_integral/fv`  |       0.93 |       1.47 | +58.28% |
| `meshgenerators/elements_to_simplices_converter.test/simple_prism_convert`      |       0.81 |       1.27 | +57.14% |
| `meshgenerators/refine_sideset_generator.generate/both`                         |       0.85 |       1.33 | +55.81% |
| `materials/coupled_value_function.order_shuffled`                               |       0.93 |       1.45 | +54.99% |
| `outputs/debug.show_execution_order/fv_interface_kernels`                       |       0.82 |       1.25 | +52.99% |
| `fvkernels/boundary_execution.force_external_boundary`                          |       0.9  |       1.38 | +52.68% |
| `convergence/reference_residual_convergence.wildcard`                           |       0.86 |       1.31 | +52.67% |
| `auxkernels/time_derivative_aux.variable`                                       |       0.95 |       1.42 | +50.47% |
## Modules test summary

### Removed tests
None
### Added tests
| Test                                                                                                                        | Time (s)   |
|-----------------------------------------------------------------------------------------------------------------------------|------------|
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_heated_physics`          | 4.88       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_scalar_physics`          | 3.65       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_physics`                 | 3.64       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_heated_enthalpy_physics` |            |
### Run time changes
| Test                                                      |   Base (s) |   Head (s) | +/-     |
|-----------------------------------------------------------|------------|------------|---------|
| `subchannel/test:mesh/detailed_quad_sub_channel_mesh.3x1` |       0.99 |       1.61 | +62.01% |

grmnptr commented on modules/navier_stokes/test/tests/finite_volume/ins/turbulence/channel/linear-segregated/tests:
```diff
@@ -11,13 +11,67 @@
+    input = 'channel_ERCOFTAC-physics.i'
+    # use same gold files
+    prereq = 'channel_ERCOFTAC_linear'
+    # Physics includes deviatoric term with dynamic viscosity, regular input does not
```
No, we are good on this!

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3385526/), step Clang format on 0063985 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31229/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31229/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b79a48d659f36efef4240a0fa85dd5bd76ac8927`

GiudGiud committed:
Address restart issues in nonlinear k-epsilon due to changes:
- redefine the turbulent viscosity

GiudGiud committed:
Adapt tests for missing 'rho_h' postprocessed quantity
Add PP for scalar conservation

GiudGiud committed:
More comments on what we are doing for heat transfer funcntor materials
Add functor material as soon as we know the FPs

GiudGiud committed:
Dont forget the scaling factor on dirichlet boundaries in turbulent diffusion

GiudGiud committed:
Regold channel test for linear + kepsilon
Add a test to capture the added coefficient and make sure
the new functor materials do the work

GiudGiud committed:
Make Physics test consistent with regular kernel syntax

GiudGiud committed:
Make C_mu required: would error otherwise

GiudGiud committed:
Add an option to specify the RC UO name, for convenience when mimicking a kernel input

GiudGiud committed:
Use FV ICs for performance and to match closer to regular FV inputs

GiudGiud committed:
Simplify inputs
more in common postprocessing and executioner includes

GiudGiud committed:
Fix deviatoric term in ERCOFTAC input (for regular mu)
Regold

GiudGiud committed:
Regold post fix in turbulent diffusion kernels
Add some prereq to facilitate debugging k-epsilon

GiudGiud committed:
Add warnings to linear FV flux kernel classes on doubling the BC application
Add notes in linearFV physics code
Move scalar diffusion terms out of the turbulence physics for linearFV

Regold physics + scalar kernel tests due to diff change

GiudGiud committed:
Regold everything due to fix in turbulent diffusion kernels

GiudGiud committed:
Add tasks dedicated for physics adding variables, ICs and materials, so that
Physics can decide what to do based on existing variables, ICs and materials

GiudGiud committed:
Clean turbulent thermal conductivity creation

GiudGiud committed:
Only add the turbulent terms when the physics actually has a turbulence model
Fixup test for defining ICs and ignoring them (error)

GiudGiud committed:
Add more required tasks
Adapt all physics to new set physics-specific tasks

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3387450/), step Results summary on fed6ac0 wanted to post the following:

## Framework test summary

Compared against b79a48d in job [civet.inl.gov/job/3380566](https://civet.inl.gov/job/3380566).

No change
## Modules test summary

Compared against b79a48d in job [civet.inl.gov/job/3380566](https://civet.inl.gov/job/3380566).

### Removed tests

| Test                                                                                                   | Time (s)   |
|--------------------------------------------------------------------------------------------------------|------------|
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC`           | 4.15       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_scaled_Re` | 4.02       |

### Added tests

| Test                                                                                                                        | Time (s)   |
|-----------------------------------------------------------------------------------------------------------------------------|------------|
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_heated_physics`          | 4.67       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_heated_enthalpy_physics` | 4.40       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear`                         | 3.83       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_scaled_Re`               | 3.76       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_scalar_physics`          | 3.65       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_lumped_eff`              | 3.54       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_physics`                 | 3.50       |

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3387894/), step Results summary on 8092bd5 wanted to post the following:

## Framework test summary

Compared against b79a48d in job [civet.inl.gov/job/3380566](https://civet.inl.gov/job/3380566).

No change
## Modules test summary

Compared against b79a48d in job [civet.inl.gov/job/3380566](https://civet.inl.gov/job/3380566).

### Removed tests

| Test                                                                                                   | Time (s)   |
|--------------------------------------------------------------------------------------------------------|------------|
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC`           | 4.15       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_scaled_Re` | 4.02       |

### Added tests

| Test                                                                                                                        | Time (s)   |
|-----------------------------------------------------------------------------------------------------------------------------|------------|
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_heated_physics`          | 4.83       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_heated_enthalpy_physics` | 4.58       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear`                         | 3.95       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_scaled_Re`               | 3.86       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_scalar_physics`          | 3.84       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_physics`                 | 3.68       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_lumped_eff`              | 3.49       |

GiudGiud committed:
Address test failures: double ; in register base physics task

grmnptr commented on modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C:
```diff
@@ -0,0 +1,324 @@
+}
+
+WCNSLinearFVTurbulencePhysics::WCNSLinearFVTurbulencePhysics(const InputParameters & parameters)
+  : WCNSFVTurbulencePhysicsBase(parameters)
```
At some point it would be nice to create different physics for the different flavors of turbulence. MixingLenghtTurbulencePhysics, KEpsilonturbulence physics etc. Maybe not for now just sometime in the future.

GiudGiud replied:
I was thinking we might get away with a single class. It works fine for WCNSFV. I think having a second model might be the test of that

grmnptr commented on modules/navier_stokes/src/physics/WCNSLinearFVTurbulencePhysics.C:
```diff
@@ -0,0 +1,324 @@
+    params.set<MooseFunctorName>(NS::mu_t + "_inverse_factor") =
+        getParam<MooseFunctorName>("sigma_eps");
+    getProblem().addMaterial(mat_type, prefix() + "mu_eff_tked", params);
+  }
```
```suggestion
  // Functor materials for the diffusion coefficients
  if (_turbulence_model == "k-epsilon")
  {
    // Since sigma_k = 1 in the standard k-epsilon, this is often unnecessary
    const std::string mat_type = "FunctorEffectiveDynamicViscosity";
    InputParameters params = getFactory().getValidParams(mat_type);
    assignBlocks(params, _blocks);
    params.set<MooseFunctorName>("property_name") = "mu_eff_tke";
    params.set<MooseFunctorName>(NS::mu) = _flow_equations_physics->dynamicViscosityName();
    params.set<MooseFunctorName>(NS::mu_t) = _turbulent_viscosity_name;
    params.set<MooseFunctorName>(NS::mu_t + "_inverse_factor") =
        getParam<MooseFunctorName>("sigma_k");
    getProblem().addMaterial(mat_type, prefix() + "mu_eff_tke", params);
  
    params.set<MooseFunctorName>("property_name") = "mu_eff_tked";
    params.set<MooseFunctorName>(NS::mu_t + "_inverse_factor") =
        getParam<MooseFunctorName>("sigma_eps");
    getProblem().addMaterial(mat_type, prefix() + "mu_eff_tked", params);
  }
```

GiudGiud committed:
Address review: less code for effective viscosities

Co-authored-by: Peter German <31662443+grmnptr@users.noreply.github.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3396408/), step Clang format on 553b55a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31229/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31229/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format fd68302d17269526eecef8e3f4924f2b825f3911`

grmnptr commented on modules/navier_stokes/src/physics/WCNSLinearFVFluidHeatTransferPhysics.C:
```diff
@@ -139,13 +139,15 @@ WCNSLinearFVFluidHeatTransferPhysics::addEnergyHeatConductionKernels()
-          _thermal_conductivity_name[block_i] + "_by_cp";
+      const auto th_cond_name =
+          _thermal_conductivity_name[block_i] + (_has_turbulence_model ? "_plus_kt" : "");
+      params.set<MooseFunctorName>("diffusion_coeff") = th_cond_name + "_by_cp";
```
https://en.wikipedia.org/wiki/Thermal_diffusivity

grmnptr commented on modules/navier_stokes/src/linearfvkernels/LinearFVTurbulentDiffusion.C:
```diff
@@ -89,7 +89,8 @@ LinearFVTurbulentDiffusion::computeBoundaryMatrixContribution(const LinearFVBoun
   {
     const auto face_arg = singleSidedFaceArg(_current_face_info);
-    grad_contrib *= _diffusion_coeff(face_arg, determineState());
+    grad_contrib *=
```
Okay I'll do it then.

grmnptr reviewed (approved):
Looks good from my side. 
Will get someone for a technical review.

tanoret committed:
adding consistent projection (SIMPLEC) to flow physics - needed for bfs models

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3423868/), step Results summary on 96866ce wanted to post the following:

## Framework test summary

Compared against 82bb60c in job [civet.inl.gov/job/3419953](https://civet.inl.gov/job/3419953).

No change
## Modules test summary

Compared against 82bb60c in job [civet.inl.gov/job/3419953](https://civet.inl.gov/job/3419953).

### Removed tests

| Test                                                                                                   | Time (s)   |
|--------------------------------------------------------------------------------------------------------|------------|
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC`           | 3.87       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_scaled_Re` | 3.84       |

### Added tests

| Test                                                                                                                        | Time (s)   |
|-----------------------------------------------------------------------------------------------------------------------------|------------|
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_heated_physics`          | 5.10       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_heated_enthalpy_physics` | 4.81       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_lumped_eff`              | 4.73       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_scaled_Re`               | 4.56       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_scalar_physics`          | 4.10       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear_physics`                 | 4.02       |
| `navier_stokes/test:finite_volume/ins/turbulence/channel/linear-segregated.channel_ERCOFTAC_linear`                         | 3.94       |

grmnptr commented:
Aaa sorry, I didn't see the details above. It is good to merge from my side, the PR is ready on OpenPronghorn.

GiudGiud commented:
thanks for the reviews guys!
Let's get your PRs in next Peter
